bochs/iodev/usb/usb_uasp.cc

```diff
@@ -0,0 +1,512 @@
+  { 0xA8, 0x00,   12, USB_TOKEN_IN,  U_IS_LBA,  UASP_FROM_COMMAND,  6, 4 },   // READ_12
+  { 0x88, 0x00,   16, USB_TOKEN_IN,  U_IS_LBA,  UASP_FROM_COMMAND, 10, 4 },   // READ_16
+  { 0x0A, 0x00,    6, USB_TOKEN_OUT, U_IS_LBA,  UASP_FROM_COMMAND,  2, 2 },   // WRITE_6
+  { 0x2A, 0x00,   10, USB_TOKEN_OUT, U_IS_LBA,  UASP_FROM_COMMAND,  7, 2 },   // WRITE_10
+  { 0xAA, 0x00,   12, USB_TOKEN_OUT, U_IS_LBA,  UASP_FROM_COMMAND,  6, 4 },   // WRITE_12
+  { 0x8A, 0x00,   16, USB_TOKEN_IN,  U_IS_LBA,  UASP_FROM_COMMAND, 10, 4 },   // WRITE_16
+  { 0x1E, 0x00,    6, 0,             U_NONE,                    0,       },   // PREVENT_ALLOW_MEDIUM_REMOVAL
+  { 0x25, 0x00,   10, USB_TOKEN_IN,  U_NONE,                    8,       },   // READ_CAPACITY_10
+  { 0x9E, 0x10,   16, USB_TOKEN_IN,  U_SRV_ACT, UASP_FROM_COMMAND, 10, 4 },   // READ_CAPACITY_16
+  { 0x1B, 0x00,    6, 0,             U_NONE,                    0,       },   // START_STOP_UNIT
+  { 0x2F, 0x00,   10, 0,             U_NONE,                    0,       },   // VERIFY_10
+  { 0xAF, 0x00,   12, 0,             U_NONE,                    0,       },   // VERIFY_12
+  { 0x8F, 0x00,   16, 0,             U_NONE,                    0,       },   // VERIFY_16
+  { 0x04, 0x00,    6, 0,             U_NONE,                    0,       },   // FORMAT_UNIT
+  { 0x35, 0x00,   10, 0,             U_NONE,                    0,       },   // SYNCHRONIZE CACHE_10
+  { 0x91, 0x00,   16, 0,             U_NONE,                    0,       },   // SYNCHRONIZE CACHE_16
+// end marker
+  { 0xFF, }
+};
+
+Bit32u usb_msd_device_c::get_data_len(const struct S_UASP_INPUT *input, Bit8u *buf)
+{
+  Bit32u ret = 0;
+
+  // the input offset and size fields point to the commands "data_len" field
+  switch (input->size) {
+    case 1: // byte
+      ret = * (Bit8u *) &buf[input->offset];
+      break;
+    case 2: // word
+      ret = * (Bit16u *) &buf[input->offset];
+      ret = bx_bswap16(ret);
+      break;
+    case 4: // dword
+      ret = * (Bit32u *) &buf[input->offset];
+      ret = bx_bswap32(ret);
+      break;
+  }
+
+  // is lba instead of bytes?
+  if (input->flags & U_IS_LBA)
+    ret *= 512;  // TODO: get size of a sector/block
+
+  return ret;
+}
+
+// get information about the command requested
+const struct S_UASP_INPUT *usb_msd_device_c::uasp_get_info(Bit8u command, Bit8u serv_action)
+{
+  int i = 0;
+
+  do {
+    if (bx_uasp_info_array[i].command == command) {
+      if (bx_uasp_info_array[i].flags & U_SRV_ACT) {
+        if (bx_uasp_info_array[i].serv_action == serv_action)
+          return &bx_uasp_info_array[i];
+      } else
+        return &bx_uasp_info_array[i];
+    }
+    i++;
+  } while (bx_uasp_info_array[i].command != 0xFF);
+
+  BX_ERROR(("uasp_get_info: Unknown command found: %02X/%02X", command, serv_action));
+  return NULL;
+}
+
+int usb_msd_device_c::uasp_do_stall(UASPRequest *req)
+{
+  USBPacket *p;
```

Comment:
better to initialize p in place

`USBPacket *p = req->p;`

Reply:
done, though does it really matter. A good compiler will do it anyway, yes?  :-)

Reply:
It is not about compiler it is about initializing the variable together with definition :)